streaming/src/main/scala/org/apache/spark/streaming/util/RecurringTimer.scala

```diff
@@ -87,22 +87,32 @@ class RecurringTimer(clock: Clock, period: Long, callback: (Long) => Unit, name:
     prevTime
   }
 
+  private def triggerActionForNextInterval(): Unit = {
+    clock.waitTillTime(nextTime)
+    callback(nextTime)
+    prevTime = nextTime
+    nextTime += period
+    logDebug("Callback for " + name + " called at time " + prevTime)
+  }
+
   /**
    * Repeatedly call the callback every interval.
    */
   private def loop() {
     try {
       while (!stopped) {
-        clock.waitTillTime(nextTime)
-        callback(nextTime)
-        prevTime = nextTime
-        nextTime += period
-        logDebug("Callback for " + name + " called at time " + prevTime)
+        triggerActionForNextInterval()
       }
+      triggerActionForNextInterval()
     } catch {
       case e: InterruptedException =>
     }
   }
+
+  /*
+   * Return whether `stop` is called.
+   */
+  def isStopped: Boolean = stopped
```

Comment:
To be clear this does not signify that stopping is over (that is calling the callback has been stopped). This is just signifies that the stopping processing has been started.

I am wondering whether this is ambiguous, and it may be cleaner to actually capture the three states - Initialized, Started, Stopping, Stopped. Similar to what you did for the ReceiverTracker some time ago. Then it becomes easy to read, reason about and test.